servers/zms/src/test/java/com/yahoo/athenz/zms/ZMSServiceTagsTest.java

```diff
@@ -0,0 +1,273 @@
+            zmsImpl.putServiceIdentity(ctx, domainName, serviceWithTags, auditRef, false, service);
+
+            // put service with single tags
+            final String serviceSingleTag = "serviceSingleTag";
+            List<String> singleTagValue = Collections.singletonList("val1");
+            service = zmsTestInitializer.createServiceObject(domainName, serviceSingleTag, "http://localhost", "/usr/bin/java", "root",
+                    "users", "host1");
+            service.setTags(Collections.singletonMap(tagKey, new TagValueList().setList(singleTagValue)));
+            zmsImpl.putServiceIdentity(ctx, domainName, serviceSingleTag, auditRef, false, service);
+
+            //put service without tags
+            final String noTagsService = "noTagsService";
+            service = zmsTestInitializer.createServiceObject(domainName, noTagsService, "http://localhost", "/usr/bin/java", "root",
+                    "users", "host1");
+            zmsImpl.putServiceIdentity(ctx, domainName, noTagsService, auditRef, false, service);
+
+            // get services without tags query - both tags should be presented
+            ServiceIdentities serviceList = zmsImpl.getServiceIdentities(ctx, domainName, Boolean.TRUE, null,null, null);
+            hasServiceWithTags(serviceList, serviceWithTags, tagKey, tagValues, 2);
+            hasServiceWithTags(serviceList, serviceSingleTag, tagKey, singleTagValue, 1);
+            hasServiceWithTags(serviceList, noTagsService, null, null, 0);
+
+            // get services with exact tag value
+            serviceList = zmsImpl.getServiceIdentities(ctx, domainName, Boolean.TRUE, null,tagKey, "val1");
+            hasServiceWithTags(serviceList, serviceWithTags, tagKey, tagValues, 2);
+            hasServiceWithTags(serviceList, serviceSingleTag, tagKey, singleTagValue, 1);
+            // ensure there are no more services
+            assertEquals(serviceList.getList().size(), 2);
+
+            // get services with exact tag value
+            serviceList = zmsImpl.getServiceIdentities(ctx, domainName, Boolean.TRUE, null, tagKey, "val2");
+            hasServiceWithTags(serviceList, serviceWithTags, tagKey, tagValues, 2);
+            // ensure there are no more services
+            assertEquals(serviceList.getList().size(), 1);
+
+            // get services with only tag key
+            serviceList = zmsImpl.getServiceIdentities(ctx, domainName, Boolean.TRUE, null, tagKey, null);
+            hasServiceWithTags(serviceList, serviceWithTags, tagKey, tagValues, 2);
+            hasServiceWithTags(serviceList, serviceSingleTag, tagKey, singleTagValue, 1);
+            // ensure there are no more services
+            assertEquals(serviceList.getList().size(), 2);
+        }
+
+        public void testProcessUpdateServiceTags() {
+            ObjectStoreConnection conn = Mockito.mock(ObjectStoreConnection.class);
+
+        }
```

Comment:
?

Reply:
by mistake create another test function that "testQueryUpdateServiceWithTags"